A31-xds-routeaction-support.md

```diff
@@ -0,0 +1,214 @@
+Balancing Policy were to queue RPCs in this scenario, wait-for-ready RPCs would
+be stuck indefinitely referencing the cluster, and it could never be removed.
+Deleted clusters need to drop the RPC and return an UNAVAILABLE status.
+
+When new clusters are added to xDS, they will be added to both the Config
+Selector and Service Config simultaneously.  Here, precautions must be taken to
+ensure it is not possible for the Config Selector to be used before the Load
+Balancing Policy's Picker has been updated to be aware of the new cluster.  The
+channel must update the Load Balancing Policy before applying the Config
+Selector, and the Load Balancing Policy must provide a new Picker before its
+Update method returns back to the channel. If the Picker does not recognize a
+cluster name provided to it, which should not be possible when following the
+steps above in that order, it will drop the RPC with an INTERNAL status code.
+
+The following drawing shows which parts of the system will contain which current
+Load Balancing Policies.  The blue box represents the Config Selector and the
+red box represents the top-level Load Balancing Policy and its children.
+
+![xDS Architecture](A31_graphics/xds_architecture.png "xDS Architecture")
+
+#### Supported Fields
+
+Initially, only timeout-related `RouteAction` settings will be supported:
+[`timeout`](https://github.com/envoyproxy/envoy/blob/7abb0e0bbed4f6b6304403b93762614ad385f80d/api/envoy/api/v2/route/route_components.proto#L886)
+and
+[`max_grpc_timeout`](https://github.com/envoyproxy/envoy/blob/7abb0e0bbed4f6b6304403b93762614ad385f80d/api/envoy/api/v2/route/route_components.proto#L979).
+gRPC's xDS support will completely ignore
+[`grpc_timeout_offset`](https://github.com/envoyproxy/envoy/blob/7abb0e0bbed4f6b6304403b93762614ad385f80d/api/envoy/api/v2/route/route_components.proto#L988).
+This field is used in Envoy for shortening timeouts set by clients to ensure the
+timeout is reached before the gRPC client cancels the RPC due to its deadline,
+so supporting it directly in the client is unnecessary and would be
+counter-productive.
+
+If `max_grpc_timeout` is not present, the `timeout` field will specify the
+maximum RPC timeout for this route, with a default of 15s.
+
+If `max_grpc_timeout` is present, then `timeout` is ignored and
+`max_grpc_timeout` limits the maximum timeout for RPCs on this route.  A value
+of 0 indicates no limit should be applied.  In all cases, the RPC timeout set by
+the application may never be exceeded due to these settings.  For examples, see
+the following table:
+
+Application Deadline | `max_grpc_timeout` | `timeout` | Effective Timeout
+-------------------- | ------------------ | --------- | -----------------
+unset | unset | unset | 15s
+unset | unset | 10s | 10s
+unset | 0s | any | infinite
+unset | 10s | any | 10s
+10s | unset | unset | 10s
+20s | unset | unset | 15s
+20s | unset | 10s | 10s
+20s | 0s | any | 20s
+20s | 10s | any | 10s
+
```

Comment:
[Envoy's API doc](https://github.com/envoyproxy/envoy/blob/c694e470dbc9a1dd24b296d117330f0e5becb2c9/api/envoy/api/v2/route/route_components.proto#L877-L878) also mentions that a value of 0 for `timeout` will disable the route's timeout. So may worthwhile to list that case as well.

Reply:
+1